akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -699,18 +829,71 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
-  final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
+  final protected def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
     new AsyncCallback[T] {
       override def invoke(event: T): Unit =
         interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
     }
   }
 
+  // TODO rethink if it's worth it; move the Updater to an object; etc
+  private final val _setStageActorRef = AtomicReferenceFieldUpdater.newUpdater(classOf[GraphStageLogic], classOf[StageActorRef], "_stageActorRefDoNotCallMeDirectly")
+  @volatile private var _stageActorRefDoNotCallMeDirectly: StageActorRef = _ // DO NOT TOUCH!
```

Comment:
alternative would be to go AtomicReference around it to be safe.
Or we don't care and keep it a normal var, and if users screw up it may throw the exception. 

Now that you mention it Roland seemed to prefer that as well... Shall I remove all volatile/atomic from here?